src/Xamarin.Android.Build.Tasks/Tasks/AndroidDotnetToolTask.cs

```diff
@@ -79,8 +80,16 @@ public override bool Execute ()
 
 		protected override string GenerateFullPathToTool ()
 		{
-			if (NeedsDotnet)
-				return "dotnet";
+			if (NeedsDotnet) {
+				var dotnetPath = Environment.GetEnvironmentVariable ("DOTNET_HOST_PATH");
+				if (!File.Exists (dotnetPath)) {
+					var processPath = Process.GetCurrentProcess().MainModule.FileName;
+					if (Path.GetFileNameWithoutExtension (processPath) == "dotnet") {
+						dotnetPath = processPath;
+					}
+				}
+				return File.Exists (dotnetPath) ? dotnetPath : "dotnet";
```

Comment:
Maybe this should be a task parameter that is passed in instead? It looks like the `$(NetCoreRoot)` msbuild property could also be used for this.

Reply:
The `Microsoft.NET.ILLink.targets` snippet confuses me (aside: what *doesn't* confuse me today?), because it's setting various properties when `$(DOTNET_HOST_PATH)` *isn't* set.  But what happens when `$(DOTNET_HOST_PATH)` *is* set?

```xml
   <!-- When running from Desktop MSBuild, DOTNET_HOST_PATH is not set.
         In this case, explicitly specify the path to the dotnet host. -->
    <PropertyGroup Condition=" '$(DOTNET_HOST_PATH)' == '' ">
      <_DotNetHostDirectory>$(NetCoreRoot)</_DotNetHostDirectory>
      <_DotNetHostFileName>dotnet</_DotNetHostFileName>
      <_DotNetHostFileName Condition="$([MSBuild]::IsOSPlatform(`Windows`))">dotnet.exe</_DotNetHostFileName>
    </PropertyGroup>
```

I ask because *the above properties* are used elsewhere, e.g. https://github.com/dotnet/sdk/blob/b15a88fb330548bf447d2df4936c4367db346a1a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets#L185-L186

```xml
    <ILLink AssemblyPaths="@(ManagedAssemblyToLink)"
            …
            ToolExe="$(_DotNetHostFileName)"
            ToolPath="$(_DotNetHostDirectory)"
```

So when `$(DOTNET_HOST_PATH)` *is* set, those properties *won't* be set, so… shouldn't it just fail horrifically, as `ToolExe` and `ToolPath` are the empty string?

Reply:
The answer to my confusion appears to be that the `<ILLink/>` task *doesn't use* `ToolExe` and `ToolPath`:

https://github.com/dotnet/linker/blob/87539d443b197b63b99fabf603e20811bad632c5/src/ILLink.Tasks/LinkTask.cs#L224-L239

```csharp
	partial class ILLink {
		private const string DotNetHostPathEnvironmentName = "DOTNET_HOST_PATH";

		private string _dotnetPath;

		private string DotNetPath {
			get {
				if (!String.IsNullOrEmpty (_dotnetPath))
					return _dotnetPath;

				_dotnetPath = Environment.GetEnvironmentVariable (DotNetHostPathEnvironmentName);
				if (String.IsNullOrEmpty (_dotnetPath))
					throw new InvalidOperationException ($"{DotNetHostPathEnvironmentName} is not set");

				return _dotnetPath;
			}
		}
		// …
		protected override string GenerateFullPathToTool () => DotNetPath;
	}
```

which is *differently* "horrifying"; *this always throws* if `$(DOTNET_HOST_PATH)` isn't set, and *never* uses `ToolExe` & `ToolPath`, so… How does this work when `$(DOTNET_HOST_PATH)` *isn't* set?

The confusion compounds.

Reply:
By my understanding `GenerateFullPathToTool ()` is only called when `ToolPath` and `ToolExe` do not point to a valid tool: https://github.com/dotnet/msbuild/blob/f1dae6ab690483458d37b8900f1d1e4a5fc72851/src/Utilities/ToolTask.cs#L463-L481

I think what effectively happens here is that `$(_DotNetHostFileName)` and `$(_DotNetHostDirectory)` are preferred and used if valid, and `$(DOTNET_HOST_PATH)` is used as a fallback if not.